src/cycle_basis.c

```diff
@@ -0,0 +1,217 @@
+/* -*- mode: C -*-  */
+/* 
+   IGraph library.
+   Copyright (C) 2010-2012  Gabor Csardi <csardi.gabor@gmail.com>
+   334 Harvard street, Cambridge, MA 02139 USA
+   Copyright (C) 2019 the igraph team.
+   
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 
+   02110-1301 USA
+
+*/
+
+#include "igraph_structural.h"
+#include "igraph_dqueue.h"
+#include "igraph_interface.h"
+#include "igraph_interrupt_internal.h"
+#include "igraph_memory.h"
+#include "igraph_types_internal.h"
+#include "igraph_visitor.h"
+
+/**
+ * \function igraph_cycle_basis_unweighted_undirected
+ * Finds a cycle basis for an unweighted, undirected graph
+ * 
+ * A cycle basis is a set of cycles such that the set difference
+ * between them (i.e. the set of edges that are counted an odd
+ * number of times by the cycles) is a Eulerian path across the
+ * graph, which means it touches all vertices once and never
+ * walks the same edge more than once.
+ *
+ * For unweighted, undirected graphs a cycle basis is always found
+ * by calculating a spanning tree and closing it with edges (??)
+ *
+ * \param graph The input graph, it might be directed, but edge
+ *    direction is ignored.
+ * \param cycles An initialized vector of graphs. Each cycle will be
+ *    a member of this vector (as a graph).
+ * \return Error code.
+ * 
+ * Time complexity: ???.
+ * 
+ * \sa \ref igraph_maximum_cardinality_search().
+ */
+int igraph_cycle_basis_unweighted_undirected(const igraph_t* graph,
+    igraph_vector_t* cycles) {
+
+    // FIXME: deal with unconnected graphs
```

Comment:
Check whether MSVC in C89 mode is okay with comments starting with double slashes. I have always tried to avoid this because I wasn't sure whether MSVC likes it or not and it's too complicated for me to get hold of a Windows machine :)

Reply:
MSVC 2015 and later do support `//` comments as "a Microsoft extension", unless standards mode is triggered. https://docs.microsoft.com/en-us/cpp/c-language/c-comments?view=vs-2015 This can also be tested e.g. at godbolt.org

I _think_ earlier versions support it too, but I do not have proof.

I always went with `/* ... */` _just in case_. You never know what compilers people use on what weird platforms that we don't have access to. Maybe someone uses an IBM compiler on a Power architecture machine. I'd be in favour of sticking with the standard (even though there are already several `//` comments in the igraph code base—my IDE keeps flagging them actually).

Reply:
`//` is C99 I believe. Older MSVC compilers do no support it. I don't know what Python 2.x uses now to compile its extensions, but R is C99, so if you want to use them I don't mind.

Reply:
There are two separate issues:

 1. Is it okay to use arbitrary C99 features? No, because even MSVC 2019 doesn't support them fully. We should not use e.g. variable-length arrays for this reason.

 2. Is it okay to use `//` comments? Compilers typically support `//` as an extension, even if they do not fully support C99. So it might not be a problem in practice. It is claimed that [MSVC 2010 already supported `//` comments](https://softwareengineering.stackexchange.com/q/224747/29979).